File: prob7b/codes/prob7.py
import math

# Define the probability distribution for X
px = { -1: 1/4, 0: 1/2, 1: 1/4 }

# Function to calculate the entropy of a random variable
def entropy(probabilities):
    H = 0
    for p in probabilities:
        if p > 0:
            H -= p * math.log2(p)
    return H

# Calculate the entropy of X
HX = entropy(px.values())

# Calculate the entropies of the other random variables
H2X = entropy([px[x]*px[y] for x in px for y in px])
HX2 = entropy([px[x]**2 for x in px])
H2X_exp = entropy([px[x]*math.log2(2) for x in px])

# Check the statements
A = HX <= math.log2(len(px))  # log2(K) where K is the number of distinct values of X
B = HX <= H2X
C = HX <= HX2
D = HX <= H2X_exp

# Print the results
print("Statement A is :", A)
print("Statement B is :", B)
print("Statement C is :", C)
print("Statement D is :", D)

File: prob7/codes/prob7.c
#include <stdio.h>
#include <math.h>

double entropy(int K, double *probabilities) {
    double H = 0;
    for (int i = 0; i < K; i++) {
        if (probabilities[i] > 0) {
            H -= probabilities[i] * log2(probabilities[i]);
        }
    }
    return H;
}

int main() {
    // Define the number of possible values K and their probabilities for variable X
    int K = 4;
    double probabilities_X[] = {0.2, 0.3, 0.2, 0.3};

    // Calculate H(X)
    double H_X = entropy(K, probabilities_X);

    // Define probabilities for 2X, X^2, and 2^X
    double probabilities_2X[] = {0.2, 0.3, 0.2, 0.3}; // Replace with the actual values
    double probabilities_X2[] = {0.1, 0.2, 0.1, 0.6}; // Replace with the actual values
    double probabilities_2X_X[] = {0.2, 0.3, 0.2, 0.3}; // Replace with the actual values

    // Calculate the entropies for each option
    double H_2X = entropy(K, probabilities_2X);
    double H_X2 = entropy(K, probabilities_X2);
    double H_2X_X = entropy(K, probabilities_2X_X);

    // Check the statements
    if (H_X <= log2(K)) {
        printf("A. H(X) <= log2(K) is true\n");
    }
    if (H_X <= H_2X) {
        printf("B. H(X) <= H(2X) is true\n");
    }
    if (H_X <= H_X2) {
        printf("C. H(X) <= H(X2) is true\n");
    }
    if (H_X <= H_2X_X) {
        printf("D. H(X) <= H(2^X) is true\n");
    }

    return 0;
}

